dds/idl/dynamic_data_adapter_generator.cpp

```diff
@@ -450,26 +454,133 @@ namespace {
+            "  DCPS::serialized_size(enc, size, distinct_value);\n";
+        } else if (forany) { // array only (classic mapping)
+          be_global->impl_ <<
+            "  OpenDDS::DCPS::serialized_size(enc, size, " << cpp_name << "_forany(value_));\n";
+        } else {
+          be_global->impl_ <<
+            "  OpenDDS::DCPS::serialized_size(enc, size, value_);\n";
+        }
+        be_global->impl_ <<
+          "  return true;\n";
+      }
+
       be_global->impl_ <<
+        "}\n"
+        "\n"
+        "bool " << impl_classname << "::serialize(\n"
+        "  OpenDDS::DCPS::Serializer& ser, OpenDDS::DCPS::Sample::Extent ext) const\n"
+        "{\n";
+      if (struct_node || union_node) {
+        be_global->impl_ <<
+          "  using namespace OpenDDS::DCPS;\n"
+          "  if (ext == Sample::Full) {\n"
+          "    return ser << value_;\n"
+          "  } else if (ext == Sample::NestedKeyOnly) {\n"
+          "    NestedKeyOnly<const " << cpp_name << "> nested_key_only(value_);\n"
+          "    return ser << nested_key_only;\n"
+          "  } else {\n";
+        if (is_topic_type) {
+          be_global->impl_ <<
+            "    KeyOnly<const " << cpp_name << "> key_only(value_);\n"
+            "    return ser << key_only;\n";
+        } else {
+          be_global->impl_ <<
+            "    return false; // Non-topic type\n";
+        }
+        be_global->impl_ <<
+          "  }\n";
+      } else {
+        be_global->impl_ <<
+          "  ACE_UNUSED_ARG(ext);\n"
+          "  using namespace OpenDDS::DCPS;\n";
+        if (distinct_type) {
+          RefWrapper distinct_type_wrapper(dynamic_cast<AST_Type*>(node), cpp_name, "");
+          distinct_type_wrapper.done();
+          be_global->impl_ <<
+            "  " << distinct_type_wrapper.wrapped_type_name() << " distinct_value(value_);\n"
+            "  return ser << distinct_value;\n";
+        } else if (forany) {
+          be_global->impl_ <<
+            "  return ser << " << cpp_name << "_forany(value_);\n";
+        } else {
+          be_global->impl_ <<
+            "  return ser << value_;\n";
+        }
+      }
+      be_global->impl_ <<
+        "}\n"
+        "\n";
+
+      be_global->header_ <<
         "protected:\n";
 
-      if (!generate_dynamic_data_adapter_access(node, wrapper, false)) {
+      if (!generate_dynamic_data_adapter_access(node, wrapper, false, impl_classname)) {
         return false;
       }
       be_global->impl_ << "\n";
-      if (!generate_dynamic_data_adapter_access(node, wrapper, true)) {
+      if (!generate_dynamic_data_adapter_access(node, wrapper, true, impl_classname)) {
         return false;
       }
 
-      be_global->impl_ <<
+      be_global->header_ <<
         "};\n"
         "\n";
     }
 
     {
       NoSafetyProfileGuard nspg;
-
-      wrapper.generate_tag();
+      if (!generate) {
+        wrapper.generate_tag();
```

Comment:
It's seem counter-intuitive that when generate is false this calls a generate_* function.

Reply:
Refactored to merge this block with the `if (generate)` block before this. It will make this clearer.